### PR TITLE
feat(): user permissions

### PR DESCRIPTION
**Pull request by Izak88**

### CODE REVIEW

File: src/api/db/permission.ts [hunk 1]
```diff
@@ -0,0 +1,42 @@
+import { Permission } from './model';
+import { getUsers } from './user';
+
+export function getPermission(id: number): Promise<any> {
+  return new Promise((resolve, reject) => {
+    new Permission({ id: id }).fetch()
+      .then(permission => !permission ? reject(permission) : resolve(permission.toJSON()));
```
Comment (jkuri) [inline src/api/db/permission.ts]: nice way of returning `Promise`, saved a lot of space 👍

File: src/api/db/build.ts [hunk 2]
```diff
@@ -1,17 +1,23 @@
 import { Build, BuildRun, Job } from './model';
 import { getLastRun } from './job';
 
-export function getBuilds(limit: number, offset: number): Promise<any> {
+export function getBuilds(userId: number, limit: number, offset: number): Promise<any> {
```
Comment (jkuri) [inline src/api/db/build.ts]: can we make `userId?: number` optional? so admin users get all builds any way without checking the permissions later.

File: src/api/db/build.ts [hunk 3]
```diff
@@ -1,17 +1,23 @@
...
       .query(q => q.orderBy('id', 'DESC').offset(offset).limit(limit))
-      .fetchAll({ withRelated: ['repository', 'jobs.runs'] })
+      .fetchAll({ withRelated: ['repository.permissions', 'jobs.runs'] })
       .then(builds => {
         if (!builds) {
           reject();
         }
 
         builds = builds.toJSON();
+        builds = builds.filter(b => {
```
Comment (jkuri) [inline src/api/db/build.ts]: this is not okay for many reasons:
- limit and offset will not work properly
- you are fetching all results and then filtering array by permissions, so its a high possibility that you fetch data from database that you will not need. this is overkill.

what I would suggest here is simply like you did here https://github.com/Izak88/abstruse/blob/481faefd50f97667618e0660e94a5ef0a249f7dc/src/api/db/repository.ts#L100. that would be the right way.

File: src/api/db/build.ts [hunk 4]
```diff
@@ -63,6 +73,10 @@ export function getBuild(id: number): Promise<any> {
           return run;
         });
 
+        if (userId) {
+          build.repository.permissions = build.repository.permissions
```
Comment (jkuri) [inline src/api/db/build.ts]: please see https://github.com/bleenco/abstruse/pull/84/files#r133958861

File: src/api/db/job.ts [hunk 5]
```diff
@@ -1,15 +1,20 @@
...
   return new Promise((resolve, reject) => {
-    new Job({ id: jobId }).fetch({ withRelated: ['build.repository', 'runs'] })
+    new Job({ id: jobId }).fetch({ withRelated: ['build.repository.permissions', 'runs'] })
       .then(job => {
         if (!job) {
           reject();
         }
+        job = job.toJSON();
+        if (userId) {
+          job.build.repository.permissions = job.build.repository.permissions
```
Comment (jkuri) [inline src/api/db/job.ts]: please see https://github.com/bleenco/abstruse/pull/84/files#r133958861

**Comment by Izak88** [comment]
Changes after CR

**Comment by irmana** [comment]
@Izak88 how far are you with this PR? thx

File: src/api/db/build.ts [hunk 6]
```diff
@@ -1,11 +1,20 @@
...
-export function getBuilds(limit: number, offset: number): Promise<any> {
+export function getBuilds(limit: number, offset: number, userId?: number): Promise<any> {
   return new Promise((resolve, reject) => {
     new Build()
-      .query(q => q.orderBy('id', 'DESC').offset(offset).limit(limit))
-      .fetchAll({ withRelated: ['repository', 'jobs.runs'] })
+      .query(q => {
+        q.innerJoin('repositories', 'repositories.id', 'builds.repositories_id')
+        .innerJoin('permissions', 'permissions.repositories_id', 'repositories.id')
+        .where('permissions.users_id', userId)
```
Comment (jkuri) [inline src/api/db/build.ts]: can you make this conditional, what if we don't pass `userId` in?

File: src/api/db/build.ts [hunk 7]
```diff
@@ -31,10 +40,21 @@ export function getBuilds(limit: number, offset: number): Promise<any> {
...
+            if (userId) {
+              query.where('permissions.users_id', userId)
+              .andWhere('permissions.permission', true)
+              .orWhere('private', false);
+            }
+          }
+        },
+        'repository.access_token',
+        'jobs.runs',
+        'runs.job_runs']})
```
Comment (jkuri) [inline src/api/db/build.ts]: style: `})` in newline

File: src/api/server-routes.ts [hunk 8]
```diff
@@ -51,19 +51,19 @@ export function buildRoutes(): express.Router {
   const router = express.Router();
 
   router.get('/limit/:limit/offset/:offset', (req: express.Request, res: express.Response) => {
```
Comment (jkuri) [inline src/api/server-routes.ts]: I don't see `userId` here in parameters. should be conditional too.

File: src/api/server-routes.ts [hunk 9]
```diff
@@ -51,19 +51,19 @@ export function buildRoutes(): express.Router {
...
-    getBuilds(req.params.limit, req.params.offset).then(builds => {
+    getBuilds(req.params.limit, req.params.offset, req.query.userId).then(builds => {
       return res.status(200).json({ data: builds });
     });
   });
 
   router.get('/last', (req: express.Request, res: express.Response) => {
```
Comment (jkuri) [inline src/api/server-routes.ts]: please make nicer URLs, with conditional parameter `userId`. query parameters are not so nice.

File: src/app/components/app-build-details/app-build-details.component.html [hunk 10]
```diff
@@ -24,7 +24,7 @@ <h1 class="bold">
                       </span>
                     </h1>
                   </div>
-                  <div class="column is-2">
+                  <div class="column is-2" *ngIf="build?.repository?.permissions[0]?.permission">
```
Comment (jkuri) [inline src/app/components/app-build-details/app-build-details.component.html]: this is very rough written `build?.repository?.permissions[0]?.permission`, I mean the `permissions[0]?.permission` part... can you reduce that in code to just `hasPermission` `boolean` value?

File: src/api/server-routes.ts (hunk 8 above)
Comment (Izak88) [inline src/api/server-routes.ts]: I don't know what you mean here, `userId` is in next line and is passed through query parameters, not directly by url...

**Pull request merged**